README.md
# raritygems

[![CI and CD](https://github.com/jojoee/raritygems/actions/workflows/continuous-integration.yml/badge.svg?branch=master)](https://github.com/jojoee/raritygems/actions/workflows/continuous-integration.yml)
[![PyPI version fury.io](https://badge.fury.io/py/raritygems.svg)](https://pypi.python.org/pypi/raritygems/)
[![codecov](https://codecov.io/gh/jojoee/raritygems/branch/master/graph/badge.svg)](https://codecov.io/gh/jojoee/raritygems)

[Provably Rare Gem Raritygems](https://gems.alphafinance.io/#/rarity) API, currently focused on mining.

## Run on Google Colab

```bash
# 1. Install packages
# !pip install --force-reinstall jsonschema==3.2.0
!python --version
!python -m pip --version
!pip install web3==5.23.0
!pip install raritygems==0.5.0

# 2. Download "raritygems_salt_finder" and give permission to execute
!wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=10K-ZTxj14_BrY_ZxupvOkGXYl2Eo9mEq' -O raritygems_salt_finder
!chmod 755 raritygems_salt_finder
```

```python
# 3. Run with Python
from raritygems import raritygems

david = raritygems.Miner(
    user_address="0xxxxx",
    gem_kind=0,  # 0 = Turquoise, 1 = Pearl, ...
    salt_finder_path="/content/raritygems_salt_finder",
    private_key="",  # if "" is provided then it will only find a salt then exit
    line_token="",  # if "" is provided then it will not notify you via LINE
)
david.mine()
```

note
- If you got "jsonschema" error then click "Runtime > Restart and run all".
- "raritygems_salt_finder" is built by golang code that can be found here https://github.com/jojoee/raritygems/blob/master/main.go
- You can check more about gem kind [here](https://github.com/jojoee/raritygems/blob/master/raritygems/helper/config.py)

## How it works?

It basically finds a "salt" and sign a transaction for you. The program will do the following

1. Get the information that needed for mining: `user_nonce` and `gem_difficulty`
2. Use [raritygems_salt_finder](https://github.com/jojoee/raritygems#raritygems_salt_finder) to find a salt
3. Sign a transaction to claim a gem
4. Go back to step 1.

## raritygems_salt_finder

At first, I start with Python and try to optimize it but I realize Go is must faster.
So, I write this "raritygems_salt_finder" part with Go language instead, and this is command I used to built it.

```
# to build "raritygems_salt_finder"
go build -o raritygems_salt_finder main.go
go build -o raritygems_salt_finder-0.5.0 main.go

# to build "raritygems_salt_finder" for Google Colab
GOOS=linux GOARCH=amd64 go build -o raritygems_salt_finder_linux main.go
GOOS=linux GOARCH=amd64 go build -o raritygems_salt_finder_linux-0.5.0 main.go
```

## Feature

- [ ] mining: auto update `gem_difficulty`
- [ ] lint: setup linting: flask8
- [x] ci-cd: pypi: publish to pypi
- [x] ci-cd: CI & CD with GitHub Actions
- [x] ci-cd: integrate semantic release
- [ ] test: writing automated test
- [ ] perf: performance evaluation
- [ ] info: display your current balance
- [ ] info: display your gems
- [ ] info: estimate how many times you can mine (before your have no gas left)
- [ ] gas: optimize gas price
- [ ] parallel: make it support parallel / multi-threading
- [ ] gpu: create GPU support version
- [ ] noti: refactor notification message
- [ ] noti: add more channels e.g. Facebook Messenger, email
- [x] web3: auto claim gems
- [ ] network: support Ethereum chain
- [ ] network: support Fantom network
- [ ] info: bot nickname
- [ ] salt_finder: auto load by default when it is not provided

## Project setup & CMD

TODO

```
conda create --name raritygems python=3.7.11
conda activate raritygems
pip uninstall raritygems -y && pip install .
go run main.go \
    -user-nonce=1234 \
    -user-address=0x28bc92e7b7e77d348fd262fb8e29da129308fbd3 \
    -chain-id=250 \
    -gem-difficulty=535996 \
    -gem-address=0x342EbF0A5ceC4404CcFF73a40f9c30288Fc72611 \
    -gem-entropy=0x000080440000047163a56455ac4bc6b1f1b88efadf17db76e5c52c0ca594fd9b \
    -gem-kind=0 \
    -salt=41543544848 \
    -n=800000 \
    -debug=true
```

## Other projects
- [Provably-Rare-Gem-Miner](https://github.com/yoyoismee/Provably-Rare-Gem-Miner)
- [ramen](https://github.com/dmptrluke/ramen)
- [GemMiner-Go](https://github.com/TkzcM/GemMiner-Go)
- [go-gem-miner](https://github.com/sorawit/go-gem-miner)

## Ref
- [Provably Rare Gem Raritygems](https://gems.alphafinance.io/#/rarity)


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.6.0 (2021-09-20)
### Feature
* **test_hash_rate:** Add function to test hash rate of "salt_finder" ([`068b30d`](https://github.com/jojoee/raritygems/commit/068b30d6643801f892ddfadaa90d39ef8911a4e0))
* **mine:** Add parameter n to control number of iterations that will be restarted ([`5e2d1d7`](https://github.com/jojoee/raritygems/commit/5e2d1d7b35720f09fb6b640a25483557ddaab89f))

### Documentation
* **readme:** Update and replace "Usage on Google Colab" with "Run on Google Colab" section ([`83bd5b6`](https://github.com/jojoee/raritygems/commit/83bd5b671d7dbb3f68a37a8d92e72bda60f2673f))
* **readme:** Update "raritygems_salt_finder" section ([`adef60b`](https://github.com/jojoee/raritygems/commit/adef60ba902a72f3dec96cdaa2e8243590e4e933))
* **readme:** Add "Other projects" section into readme.md ([`56c7ae0`](https://github.com/jojoee/raritygems/commit/56c7ae084cc5bf0e4bf986075ff27f4a8e722c81))

## v0.5.0 (2021-09-17)
### Feature
* **performance:** And "iterations per second" text when run it in debug mode ([`19465f1`](https://github.com/jojoee/raritygems/commit/19465f14a982309d5dc5fc884057809efc9c8831))
* **restart:** The process will restart every ~2 hrs to update the gem value ([`66d9b92`](https://github.com/jojoee/raritygems/commit/66d9b92155768550063c4233bddc8b7a9c1c3bc2))
* **message:** Update notification message a bit ([`723a202`](https://github.com/jojoee/raritygems/commit/723a2027eb2d8dbfe96023e71c9ac7ad11ab5f36))

### Documentation
* **readme:** Update public "salt_finder" url and example of "salt_finder" cmd ([`fd45327`](https://github.com/jojoee/raritygems/commit/fd453276deda5b4e284a9fdbd95287488f9783c0))

## v0.4.0 (2021-09-16)
### Feature
* **privatekey:** Force exit when found a salt if not provide a private key ([`747e7f3`](https://github.com/jojoee/raritygems/commit/747e7f382d92d46399a6deb5057ff3c4532b1627))

### Documentation
* **install:** Install with latest version instead of specific version ([`e399446`](https://github.com/jojoee/raritygems/commit/e399446525eb383b53807e6f446c2f0a20c91505))
* **readme:** Init readme doc ([`a3c5b24`](https://github.com/jojoee/raritygems/commit/a3c5b245e3aab6dd46f2386a7f751a63754c7126))

## v0.3.0 (2021-09-16)
### Feature
* **autosign:** Display tx when claim a gem ([`1d5b249`](https://github.com/jojoee/raritygems/commit/1d5b24901bfab3c7142ed483abe90fe3ccd2ea4e))

## v0.2.1 (2021-09-16)
### Fix
* **helper:** ModuleNotFoundError: No module named 'helper' ([`c5c4c9f`](https://github.com/jojoee/raritygems/commit/c5c4c9f89608bd8dd26d2da127518a3944d93638))

## v0.2.0 (2021-09-16)
### Feature
* **raritygems:** Integrate with salt_finder ([`1cf9deb`](https://github.com/jojoee/raritygems/commit/1cf9deb5f6ddf507d099365505b82381ef3a220b))
* **notification:** Dont send LINE notification when line_token is not provided ([`7725548`](https://github.com/jojoee/raritygems/commit/7725548e4440fc5dd8a97e0ca68a2e8c7dd09597))
* **saltfinder:** Add version 1 of saltfinder ([`e9bf45f`](https://github.com/jojoee/raritygems/commit/e9bf45fd607052a3c16e1b428868063a2260acd8))

### Fix
* **web3:** Get_transaction_count function not found ([`a41bca6`](https://github.com/jojoee/raritygems/commit/a41bca6999c5cafabbe3016385370425aeea562e))

## v0.1.0 (2021-09-15)
### Feature
* **miner:** Verison 1 of Miner class ([`f9bf794`](https://github.com/jojoee/raritygems/commit/f9bf794c3b2b71c810fe96e046bbd0d26f1a8975))

### Fix
* **helper:** Add missing helper files ([`35a3693`](https://github.com/jojoee/raritygems/commit/35a3693dbb96d74f8cad9b0ea3e59c9712d3a9a1))

### Performance
* Increase iterations per second by using golang instead ([`17b8109`](https://github.com/jojoee/raritygems/commit/17b8109dfb5c0a0f259d9168869db97ac6440bc3))


tests/test_helper.py
import unittest


class TestTest:
    def test_normal(self):
        assert 3 > 2


if __name__ == "__main__":
    unittest.main()


tests/test_raritygems.py
import unittest


class TestTest:
    def test_normal(self):
        assert 0 < 1


if __name__ == "__main__":
    unittest.main()


main.go
package main

import (
	"flag"
	"fmt"
	"github.com/ethereum/go-ethereum/common"
	solsha3 "github.com/miguelmota/go-solidity-sha3"
	"math/big"
	"time"
)

type ConfigPtr struct {
	userNonce     *int
	userAddress   *string
	chainId       *int
	gemDifficulty *int
	gemAddress    *string
	gemEntropy    *string
	gemKind       *int
	saltStart     *string
	debug         *bool
	n             *int
}

func main() {
	// flag
	var defaultN = 25000 * 60 * 60 * 2 // !2 hrs
	var cPtr = ConfigPtr{
		flag.Int("user-nonce", 0, "User nonce"),
		flag.String("user-address", "0xxxx", "User Fantom wallet address"),
		flag.Int("chain-id", 250, "Chain ID"),
		flag.Int("gem-difficulty", 0, "Gem difficult"),
		flag.String("gem-address", "0xxxx", "Gem contract address"),
		flag.String("gem-entropy", "0xxxx", "Gem entropy"),
		flag.Int("gem-kind", 0, "Gem kind"),
		flag.String("salt", "2300000", "Starter salt"),
		flag.Bool("debug", false, "Is debug version"),
		flag.Int("n", defaultN, "max number of iterations"), // default ~2 hrs
	}
	flag.Parse()

	userNonce := big.NewInt(int64(*cPtr.userNonce))
	userAddress := common.HexToAddress(*cPtr.userAddress)
	chainId := big.NewInt(int64(*cPtr.chainId))
	gemDifficulty := big.NewInt(int64(*cPtr.gemDifficulty))
	gemAddress := common.HexToAddress(*cPtr.gemAddress)
	gemEntropy := *cPtr.gemEntropy
	gemKind := big.NewInt(int64(*cPtr.gemKind))
	salt, _ := new(big.Int).SetString(*cPtr.saltStart, 10)
	debug := *cPtr.debug
	n := int64(*cPtr.n)

	// const
	total := int64(0)
	plus := big.NewInt(1)
	max, _ := new(big.Int).SetString("115792089237316195423570985008687907853269984665640564039457584007913129639935", 10)
	target := new(big.Int).Div(max, gemDifficulty)
	start := time.Now().UnixNano()

	// debug
	if debug {
		fmt.Printf("userNonce: %d\n", userNonce)
		fmt.Printf("userAddress: %s\n", userAddress)
		fmt.Printf("chainId: %d\n", chainId)
		fmt.Printf("gemDifficulty: %d\n", gemDifficulty)
		fmt.Printf("gemAddress: %s\n", gemAddress)
		fmt.Printf("gemEntropy: %s\n", gemEntropy)
		fmt.Printf("gemKind: %d\n", gemKind)
		fmt.Printf("salt: %v\n", salt)
		fmt.Printf("n: %d\n", n)
	}

	for true {
		// init
		total += 1

		// process
		hash := solsha3.SoliditySHA3(
			[]string{"uint256", "bytes32", "address", "address", "uint256", "uint256", "uint256"},
			[]interface{}{chainId, gemEntropy, gemAddress, userAddress, gemKind, userNonce, salt},
		)

		var luck = new(big.Int).SetBytes(hash)
		if luck.Cmp(target) != 1 {
			fmt.Print(salt)
			break
		}

		// restart
		if total >= n {
			fmt.Print(-1)
			break
		}

		// debug
		if debug {
			now := time.Now().UnixNano()
			secs := (now-start)/1e9 + 1
			fmt.Printf("\rtotal hashes %d in %d secs, hashes per second : %d", total, secs, total/secs)
		}

		// next
		salt.Add(salt, plus)
	}
}


raritygems/helper/config.py
"""
Fantom
0 Turquoise
1 Pearl
2 Zircon
3 Moonstone
4 Amber
5 Spinel
"""

FTM_GEM_ABI = '[{"inputs":[],"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"account","type":"address"},{"indexed":true,"internalType":"address","name":"operator","type":"address"},{"indexed":false,"internalType":"bool","name":"approved","type":"bool"}],"name":"ApprovalForAll","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"uint256","name":"kind","type":"uint256"}],"name":"Create","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"miner","type":"address"},{"indexed":true,"internalType":"uint256","name":"kind","type":"uint256"}],"name":"Mine","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"previousOwner","type":"address"},{"indexed":true,"internalType":"address","name":"newOwner","type":"address"}],"name":"OwnershipTransferred","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"operator","type":"address"},{"indexed":true,"internalType":"address","name":"from","type":"address"},{"indexed":true,"internalType":"address","name":"to","type":"address"},{"indexed":false,"internalType":"uint256[]","name":"ids","type":"uint256[]"},{"indexed":false,"internalType":"uint256[]","name":"values","type":"uint256[]"}],"name":"TransferBatch","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"operator","type":"address"},{"indexed":true,"internalType":"address","name":"from","type":"address"},{"indexed":true,"internalType":"address","name":"to","type":"address"},{"indexed":false,"internalType":"uint256","name":"id","type":"uint256"},{"indexed":false,"internalType":"uint256","name":"value","type":"uint256"}],"name":"TransferSingle","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"internalType":"string","name":"value","type":"string"},{"indexed":true,"internalType":"uint256","name":"id","type":"uint256"}],"name":"URI","type":"event"},{"inputs":[{"internalType":"uint256[]","name":"kinds","type":"uint256[]"}],"name":"acceptManager","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"account","type":"address"},{"internalType":"uint256","name":"id","type":"uint256"}],"name":"balanceOf","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address[]","name":"accounts","type":"address[]"},{"internalType":"uint256[]","name":"ids","type":"uint256[]"}],"name":"balanceOfBatch","outputs":[{"internalType":"uint256[]","name":"","type":"uint256[]"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"kind","type":"uint256"},{"internalType":"uint256","name":"amount","type":"uint256"},{"internalType":"address","name":"to","type":"address"}],"name":"craft","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"string","name":"name","type":"string"},{"internalType":"string","name":"color","type":"string"},{"internalType":"uint256","name":"difficulty","type":"uint256"},{"internalType":"uint256","name":"gemsPerMine","type":"uint256"},{"internalType":"uint256","name":"multiplier","type":"uint256"},{"internalType":"address","name":"crafter","type":"address"},{"internalType":"address","name":"manager","type":"address"}],"name":"create","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"id","type":"uint256"}],"name":"exists","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"gemCount","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"","type":"uint256"}],"name":"gems","outputs":[{"internalType":"string","name":"name","type":"string"},{"internalType":"string","name":"color","type":"string"},{"internalType":"bytes32","name":"entropy","type":"bytes32"},{"internalType":"uint256","name":"difficulty","type":"uint256"},{"internalType":"uint256","name":"gemsPerMine","type":"uint256"},{"internalType":"uint256","name":"multiplier","type":"uint256"},{"internalType":"address","name":"crafter","type":"address"},{"internalType":"address","name":"manager","type":"address"},{"internalType":"address","name":"pendingManager","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"_maxGemCount","type":"uint256"}],"name":"initialize","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"account","type":"address"},{"internalType":"address","name":"operator","type":"address"}],"name":"isApprovedForAll","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"kind","type":"uint256"},{"internalType":"uint256","name":"salt","type":"uint256"}],"name":"luck","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"maxGemCount","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"kind","type":"uint256"},{"internalType":"uint256","name":"salt","type":"uint256"}],"name":"mine","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"name","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"nonce","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"owner","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256[]","name":"kinds","type":"uint256[]"}],"name":"renounceManager","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256[]","name":"ids","type":"uint256[]"},{"internalType":"uint256[]","name":"amounts","type":"uint256[]"},{"internalType":"bytes","name":"data","type":"bytes"}],"name":"safeBatchTransferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"id","type":"uint256"},{"internalType":"uint256","name":"amount","type":"uint256"},{"internalType":"bytes","name":"data","type":"bytes"}],"name":"safeTransferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"operator","type":"address"},{"internalType":"bool","name":"approved","type":"bool"}],"name":"setApprovalForAll","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"_maxGemCount","type":"uint256"}],"name":"setMaxGemCount","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes4","name":"interfaceId","type":"bytes4"}],"name":"supportsInterface","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"id","type":"uint256"}],"name":"totalSupply","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256[]","name":"kinds","type":"uint256[]"},{"internalType":"address","name":"to","type":"address"}],"name":"transferManager","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"_owner","type":"address"}],"name":"transferOwnership","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256[]","name":"kinds","type":"uint256[]"},{"internalType":"address","name":"crafter","type":"address"}],"name":"updateCrafter","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"kind","type":"uint256"},{"internalType":"bytes32","name":"entropy","type":"bytes32"}],"name":"updateEntropy","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"kind","type":"uint256"},{"internalType":"string","name":"name","type":"string"},{"internalType":"string","name":"color","type":"string"}],"name":"updateGemInfo","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"kind","type":"uint256"},{"internalType":"uint256","name":"difficulty","type":"uint256"},{"internalType":"uint256","name":"multiplier","type":"uint256"},{"internalType":"uint256","name":"gemsPerMine","type":"uint256"}],"name":"updateMiningData","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"kind","type":"uint256"}],"name":"uri","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"}]'


raritygems/raritygems.py
import subprocess
import random
import sys
from time import sleep
from timeit import default_timer as timer
import traceback
from web3 import Web3, contract
from raritygems.helper.config import FTM_GEM_ABI
from raritygems.helper.notification import line


class Miner:
    # required
    user_address: str
    gem_kind: int
    private_key: str
    line_token: str

    # internal
    w3: Web3
    gem_contract: contract.Contract

    # internal (chain)
    chain_id: int = 250
    gem_address: str = "0x342EbF0A5ceC4404CcFF73a40f9c30288Fc72611"  # contract address
    gem_entropy: str = "0x000080440000047163a56455ac4bc6b1f1b88efadf17db76e5c52c0ca594fd9b"
    gem_abi: str = FTM_GEM_ABI
    provider_url: str = 'https://rpc.ftm.tools'

    def __init__(
            self,
            user_address: str,
            private_key: str,
            gem_kind: int,
            line_token: str,
            salt_finder_path: str,
    ):
        # user provided
        self.user_address = user_address
        self.gem_kind = gem_kind
        self.private_key = private_key
        self.line_token = line_token
        self.salt_finder_path = salt_finder_path

        # init
        self.w3: Web3 = Web3(Web3.HTTPProvider(self.provider_url))
        self.gem_contract = self.w3.eth.contract(address=self.gem_address, abi=self.gem_abi)

    def sign_transaction(self, salt):
        td = {
            "chainId": self.chain_id,
            "from": self.user_address,
            'gasPrice': self.w3.eth.gasPrice,
            "gas": 100000,  # TODO: config gas limit
            'nonce': self.w3.eth.get_transaction_count(self.user_address),
        }
        print('building transaction %s' % str(td))
        transaction = self.gem_contract.functions.mine(self.gem_kind, salt).buildTransaction(td)
        print('sign transaction')
        signed_tx = self.w3.eth.account.sign_transaction(transaction, private_key=self.private_key)
        tx_hash = self.w3.eth.send_raw_transaction(signed_tx.rawTransaction)
        tx_receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        print('transaction', transaction)
        print('txn_hash', tx_hash)
        print('tx_receipt', tx_receipt)
        print("tx_receipt['transactionHash']", tx_receipt['transactionHash'].hex())

    def test_hash_rate(self, n: int = 100000):
        salt = random.randint(1, 2 ** 256)
        gem_difficulty: int = self.gem_contract.functions.gems(5).call()[3]  # TODO: fix
        user_nonce: int = 10000  # TODO: fix
        start = timer()
        res = subprocess.check_output([
            self.salt_finder_path,
            '-user-nonce', str(user_nonce),
            '-user-address', self.user_address,
            '-chain-id', str(self.chain_id),
            '-gem-difficulty', str(gem_difficulty),
            '-gem-address', self.gem_address,
            '-gem-entropy', self.gem_entropy,
            '-gem-kind', str(self.gem_kind),
            '-salt', str(salt),
            '-n', str(n),
        ], universal_newlines=True, stderr=subprocess.STDOUT)
        end = timer()
        s = end - start
        print(res)
        print("%d hashes in %.2f seconds (hashes per second: %.2f)" % (n, s, (n / s)))

    def mine(self, n: int = 25000 * 60 * 60 * 2):
        """
        :param n: restart every n iterations
        :return:
        """
        msg = '🕒 Start mining'
        print(msg)

        while True:
            try:
                # arrange
                salt = random.randint(1, 2 ** 256)
                gem_difficulty: int = self.gem_contract.functions.gems(self.gem_kind).call()[3]
                user_nonce: int = self.gem_contract.functions.nonce(self.user_address).call()
                d = {
                    'gem_kind': self.gem_kind,
                    'gem_difficulty': gem_difficulty,
                    'user_address': self.user_address,
                    'user_nonce': user_nonce,
                    'salt': salt,
                }
                msg = '💡start next loop %s' % str(d)
                print(msg)
                line(self.line_token, msg)

                # act
                res = subprocess.check_output([
                    self.salt_finder_path,
                    '-user-nonce', str(user_nonce),
                    '-user-address', self.user_address,
                    '-chain-id', str(self.chain_id),
                    '-gem-difficulty', str(gem_difficulty),
                    '-gem-address', self.gem_address,
                    '-gem-entropy', self.gem_entropy,
                    '-gem-kind', str(self.gem_kind),
                    '-salt', str(salt),  # starter salt
                    '-n', str(n),
                ], universal_newlines=True, stderr=subprocess.STDOUT)
                print("res", res)

                # sign
                target_salt = int(res)
                d['salt'] = target_salt
                msg = '🎉 Gem found %s' % str(d)

                if target_salt == -1:
                    # just restart or update the data
                    print('🔄 update the value')
                else:
                    line(self.line_token, msg)
                    print(msg)
                    if not self.private_key:
                        sys.exit()
                    self.sign_transaction(target_salt)

                    # TODO: get tx status before go next loop (success or fail)
                    print('you are tired for mint one sleep for 16 secs')
                    sleep(16)  # sleep 16 second before get new loop

            except Exception as e:
                line(self.line_token, '🔥🔥🔥🔥🔥🔥🔥🔥\n ERROR %s' % str(e))
                print('An exception occurred: {}'.format(e))

                print('traceback.print_exc()')
                traceback.print_exc()

                print('traceback.format_exc()')
                fes = traceback.format_exc()
                print(fes)

                print('error then go sleep for 16 secs')
                sleep(16)
